dango/lending/src/execute.rs

```diff
@@ -0,0 +1,214 @@
+pub fn execute(ctx: MutableCtx, msg: ExecuteMsg) -> anyhow::Result<Response> {
+    match msg {
+        ExecuteMsg::WhitelistDenom(denom) => whitelist_denom(ctx, denom),
+        ExecuteMsg::DelistDenom(denom) => delist_denom(ctx, denom),
+        ExecuteMsg::Deposit { recipient } => deposit(ctx, recipient),
+        ExecuteMsg::Withdraw { recipient } => withdraw(ctx, recipient),
+        ExecuteMsg::Borrow { coins } => borrow(ctx, coins),
+    }
+}
+
+pub fn whitelist_denom(ctx: MutableCtx, denom: Denom) -> anyhow::Result<Response> {
+    // Ensure only chain owner can whitelist denoms
+    ensure!(
+        ctx.sender == ctx.querier.query_config()?.owner,
+        "Only the owner can whitelist denoms"
+    );
+
+    // Ensure the denom is not already in the whitelist
+    ensure!(
+        !WHITELISTED_DENOMS.has(ctx.storage, denom.clone()),
+        "Denom already whitelisted"
+    );
+
+    // Insert the denom into the whitelist
+    WHITELISTED_DENOMS.insert(ctx.storage, denom)?;
+
+    Ok(Response::new())
+}
+
+pub fn delist_denom(ctx: MutableCtx, denom: Denom) -> anyhow::Result<Response> {
+    // Ensure only chain owner can delist denoms
+    ensure!(
+        ctx.sender == ctx.querier.query_config()?.owner,
+        "Only the owner can delist denoms"
+    );
+
+    // Ensure the denom is in the whitelist
+    ensure!(
+        WHITELISTED_DENOMS.has(ctx.storage, denom.clone()),
+        "Denom not whitelisted"
+    );
+
+    // Remove the denom from the whitelist
+    WHITELISTED_DENOMS.remove(ctx.storage, denom);
+
+    Ok(Response::new())
+}
+
+/// Ensures that the sender's account is not a margin account.
+fn ensure_sender_account_is_not_margin(ctx: &MutableCtx) -> anyhow::Result<()> {
+    let account_factory: Addr = ctx.querier.query_app_config(ACCOUNT_FACTORY_KEY)?;
+    ensure!(
+        !ctx.querier
+            .query_wasm_smart(account_factory, QueryAccountRequest {
+                address: ctx.sender,
+            })?
+            .params
+            .is_margin(),
+        "Margin accounts can't deposit or withdraw"
+    );
+    Ok(())
+}
+
+/// Ensures that the sender's account is a margin account.
+fn ensure_sender_account_is_margin(ctx: &MutableCtx) -> anyhow::Result<()> {
+    let account_factory: Addr = ctx.querier.query_app_config(ACCOUNT_FACTORY_KEY)?;
+    ensure!(
+        ctx.querier
+            .query_wasm_smart(account_factory, QueryAccountRequest {
+                address: ctx.sender,
+            })?
+            .params
+            .is_margin(),
+        "Only margin accounts can borrow and repay"
+    );
+    Ok(())
+}
+
+pub fn deposit(ctx: MutableCtx, recipient: Option<Addr>) -> anyhow::Result<Response> {
+    // Ensure margin accounts can't deposit
+    ensure_sender_account_is_not_margin(&ctx)?;
+
+    // For each deposited denom, ensure it's whitelisted and mint LP tokens.
+    let cfg = ctx.querier.query_config()?;
+    let mut msgs = vec![];
+    for coin in ctx.funds.into_iter() {
```

Comment:
`Coins` implements `IntoIterator`, so you can do

```rust
for coin in ctx.funds {
    // ...
}
```

No need to call `.into_iter()`